File: Practica2/backend-python/routes/files.py
import os
from datetime import datetime
from flask import Blueprint, request, jsonify
from config.db import get_connection
from config.s3 import get_s3_client
from middleware.auth import auth_required
from dotenv import load_dotenv

load_dotenv()

files_bp = Blueprint('files', __name__)

# GET /api/files
@files_bp.route('/', methods=['GET'])
@auth_required
def get_files():
    try:
        conn = get_connection()
        cur = conn.cursor()
        cur.execute('SELECT * FROM files WHERE user_id = %s ORDER BY created_at DESC', (request.user_id,))
        rows = cur.fetchall()
        files = [{'id': r[0], 'user_id': r[1], 'filename': r[2], 'file_type': r[3], 'file_url': r[4], 'file_size': r[5], 'created_at': str(r[6])} for r in rows]
        return jsonify(files)
    except Exception as e:
        print(e)
        return jsonify({'error': 'Error al obtener archivos'}), 500
    finally:
        cur.close()
        conn.close()

# POST /api/files/upload
@files_bp.route('/upload', methods=['POST'])
@auth_required
def upload_file():
    if 'file' not in request.files:
        return jsonify({'error': 'Archivo requerido'}), 400

    file = request.files['file']
    if not file.filename:
        return jsonify({'error': 'Archivo requerido'}), 400

    try:
        filename = file.filename
        content_type = file.content_type
        key = f"files/{request.user_id}/{int(datetime.utcnow().timestamp())}_{filename}"

        file_type = 'other'
        if content_type.startswith('image/'):
            file_type = 'image'
        elif content_type.startswith('text/'):
            file_type = 'text'

        file_data = file.read()
        file_size = len(file_data)

        s3 = get_s3_client()
        import io
        s3.upload_fileobj(
            io.BytesIO(file_data),
            os.getenv('S3_BUCKET_FILES'),
            key,
            ExtraArgs={'ContentType': content_type}
        )

        file_url = f"https://{os.getenv('S3_BUCKET_FILES')}.s3.{os.getenv('AWS_REGION')}.amazonaws.com/{key}"

        conn = get_connection()
        cur = conn.cursor()
        cur.execute(
            'INSERT INTO files (user_id, filename, file_type, file_url, file_size) VALUES (%s, %s, %s, %s, %s) RETURNING *',
            (request.user_id, filename, file_type, file_url, file_size)
        )
        row = cur.fetchone()
        conn.commit()

        return jsonify({'id': row[0], 'filename': row[2], 'file_type': row[3], 'file_url': row[4], 'file_size': row[5]}), 201

    except Exception as e:
        print(e)
        return jsonify({'error': 'Error al subir archivo'}), 500
    finally:
        cur.close()
        conn.close()

# POST /api/files/upload-url — guardar URL de archivo subido por Lambda
@files_bp.route('/upload-url', methods=['POST'])
@auth_required
def upload_url():
    data = request.get_json()
    filename = data.get('filename')
    file_type = data.get('file_type')
    file_url = data.get('file_url')
    file_size = data.get('file_size')

    try:
        conn = get_connection()
        cur = conn.cursor()
        cur.execute(
            'INSERT INTO files (user_id, filename, file_type, file_url, file_size) VALUES (%s, %s, %s, %s, %s) RETURNING *',
            (request.user_id, filename, file_type, file_url, file_size)
        )
        row = cur.fetchone()
        conn.commit()
        return jsonify({'id': row[0], 'filename': row[2], 'file_type': row[3], 'file_url': row[4]}), 201
    except Exception as e:
        print(e)
        return jsonify({'error': 'Error al guardar archivo'}), 500
    finally:
        cur.close()
        conn.close()

# DELETE /api/files/<id>
@files_bp.route('/<int:file_id>', methods=['DELETE'])
@auth_required
def delete_file(file_id):
    try:
        conn = get_connection()
        cur = conn.cursor()
        cur.execute('DELETE FROM files WHERE id = %s AND user_id = %s RETURNING id', (file_id, request.user_id))
        row = cur.fetchone()
        conn.commit()
        if not row:
            return jsonify({'error': 'Archivo no encontrado'}), 404
        return jsonify({'message': 'Archivo eliminado correctamente'})
    except Exception as e:
        print(e)
        return jsonify({'error': 'Error al eliminar archivo'}), 500
    finally:
        cur.close()
        conn.close()

File: Practica2/backend-js/src/routes/files.js
const express = require('express');
const router = express.Router();
const multer = require('multer');
const { PutObjectCommand } = require('@aws-sdk/client-s3');
const pool = require('../config/db');
const s3 = require('../config/s3');
const authMiddleware = require('../middleware/auth');
require('dotenv').config();

const upload = multer({ storage: multer.memoryStorage() });

// GET /api/files — listar archivos del usuario
router.get('/', authMiddleware, async (req, res) => {
  try {
    const result = await pool.query(
      'SELECT * FROM files WHERE user_id = $1 ORDER BY created_at DESC',
      [req.userId]
    );
    res.json(result.rows);
  } catch (err) {
    console.error(err);
    res.status(500).json({ error: 'Error al obtener archivos' });
  }
});

// POST /api/files/upload — subir archivo directamente
router.post('/upload', authMiddleware, upload.single('file'), async (req, res) => {
  if (!req.file) {
    return res.status(400).json({ error: 'Archivo requerido' });
  }

  try {
    const { originalname, mimetype, buffer, size } = req.file;
    const key = `files/${req.userId}/${Date.now()}_${originalname}`;

    let file_type = 'other';
    if (mimetype.startsWith('image/')) file_type = 'image';
    else if (mimetype.startsWith('text/')) file_type = 'text';

    await s3.send(new PutObjectCommand({
      Bucket: process.env.S3_BUCKET_FILES,
      Key: key,
      Body: buffer,
      ContentType: mimetype,
    }));

    const file_url = `https://${process.env.S3_BUCKET_FILES}.s3.${process.env.AWS_REGION}.amazonaws.com/${key}`;

    const result = await pool.query(
      'INSERT INTO files (user_id, filename, file_type, file_url, file_size) VALUES ($1, $2, $3, $4, $5) RETURNING *',
      [req.userId, originalname, file_type, file_url, size]
    );

    res.status(201).json(result.rows[0]);
  } catch (err) {
    console.error(err);
    res.status(500).json({ error: 'Error al subir archivo' });
  }
});

// POST /api/files/upload-url — guardar URL de archivo subido por Lambda
router.post('/upload-url', authMiddleware, async (req, res) => {
  const { filename, file_type, file_url, file_size } = req.body;
  try {
    const result = await pool.query(
      'INSERT INTO files (user_id, filename, file_type, file_url, file_size) VALUES ($1, $2, $3, $4, $5) RETURNING *',
      [req.userId, filename, file_type, file_url, file_size]
    );
    res.status(201).json(result.rows[0]);
  } catch (err) {
    console.error(err);
    res.status(500).json({ error: 'Error al guardar archivo' });
  }
});

// DELETE /api/files/:id — eliminar archivo
router.delete('/:id', authMiddleware, async (req, res) => {
  const { id } = req.params;

  try {
    const result = await pool.query(
      'DELETE FROM files WHERE id = $1 AND user_id = $2 RETURNING id',
      [id, req.userId]
    );

    if (result.rows.length === 0) {
      return res.status(404).json({ error: 'Archivo no encontrado' });
    }

    res.json({ message: 'Archivo eliminado correctamente' });
  } catch (err) {
    console.error(err);
    res.status(500).json({ error: 'Error al eliminar archivo' });
  }
});

module.exports = router;